### PR TITLE
feat: add googlepubsub bindings

### PR DESCRIPTION
**Pull request by whitlockjc**
This commit contains two bindings, a `Channel Binding Object` and a `Message Binding Object` for Google Cloud Pub/Sub using the `googlepubsub` protocol name.

**Related issue(s)**
JSON Schema PR: https://github.com/asyncapi/spec-json-schemas/pull/253
Spec PR: https://github.com/asyncapi/spec/pull/836

### CODE REVIEW

File: googlepubsub/README.md
Comment (derberg) [inline googlepubsub/README.md:9]: you want first version to be major?

File: googlepubsub/json_schemas/channel.json
Comment (derberg) [inline googlepubsub/json_schemas/channel.json:1]: in both schemas, you're missing `bindingVersion` property. Please have a look at schemas from other bindings. You also do not have it in markdown definitions. It is necessary for user to have it, so they can specify what version of binding they're using. Bindings specs are versioned independently from main spec

File: googlepubsub/json_schemas/message.json
Comment (derberg) [inline googlepubsub/json_schemas/message.json:1]: I also noticed your schemas do not allow `patternProperties` with `x-` as the other schemas. Is it intentional, you do not want to allow extensions inside binding?

File: googlepubsub/README.md
Comment (derberg) [inline googlepubsub/README.md:70]: This is quite a workaround 😄 
Shouldn't we try to fix it on a spec level rather than add such workaround gates on a binding level?

**Review by derberg — CHANGES_REQUESTED** [review]
@whitlockjc I added few comments. Also general comment is that I'm missing examples in the markdown files, we should have them there too

Side note, to have it in the spec we will also need to add `googlepubsub` key to the spec in the list of bindings, but this is later, I can guide you and send some examples.

@fmvilas new binding on the way, please review

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:9]: I wasn't sure what to use here, just not sure what else I see changing after we agree with its implementation.  I'm all ears for suggestions.

File: googlepubsub/json_schemas/channel.json
Comment (whitlockjc) [reply googlepubsub/json_schemas/channel.json:1]: My apologies, I'll update the PR to have these in both markdown and schemas.

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:70]: Well, it's really not.  The whole purpose behind this is to describe the schema Cloud Pub/Sub uses to validate messages.  While AsyncAPI does have support for this already in the messages themselves, there is metadata on the schema object from a Cloud Pub/Sub perspective that is still useful for documenting the API.   Below is an example of how AsyncAPI's support for Avro and how its lack of Protobuf support would be documented:

```yaml
# ...
components:
    messages:
        messageAvroAsJSON:
            bindings:
                googlepubsub:
                    schema:
                        definition: |
                            {
                              "type" : "record",
                              "name" : "Message",
                              "fields" : [
                                {
                                  "name" : "message",
                                  "type" : "string"
                                }
                              ]
                            }
                        name: projects/jaydub-testing/schemas/message-avro
                        type: avro
            contentType: application/json
            name: projects/jaydub-testing/schemas/message-avro
            payload:
                fields:
                    - name: message
                      type: string
                name: Message
                type: record
            schemaFormat: application/vnd.apache.avro+yaml;version=1.9.0
        messageProto:
            bindings:
                googlepubsub:
                    schema:
                        definition: |
                            syntax = "proto2";

                            message Message {
                              required string message = 1;
                            }
                        name: projects/jaydub-testing/schemas/message-proto
                        type: protobuf
            contentType: application/octet-stream
            name: projects/jaydub-testing/schemas/message-proto
            payload: true
# ...
```

`name` is metadata that could/should be present, but I could see a case where if Protobuf were supported by AsyncAPI natively for Message payloads, maybe it wouldn't need to be present in the binding.  Thoughts?

File: googlepubsub/json_schemas/message.json
Comment (whitlockjc) [reply googlepubsub/json_schemas/message.json:1]: It was intentional at first but I really can't think of a good reason why we shouldn't.  I'll add support.

**Comment by whitlockjc** [comment]
I added examples to the Markdown file, and added `x-*` support to the JSON Schemas.

File: googlepubsub/README.md
Comment (lbroudoux) [reply googlepubsub/README.md:9]: For other bindings, first public revision is usually marked as `0.1.0`. I think we can stick with convention.

File: googlepubsub/README.md
Comment (lbroudoux) [reply googlepubsub/README.md:108]: I agree with @derberg: I think we absolutely have to avoid duplicating things in bindings when they should be defined in the message itself. Protobuf support should ideally follow the same approach as for Avro: having a dedicated `schemaFormat` entry for it. However due to `proto` syntax, we'll probably not be able to embed it in the AsyncAPI spec itself and use a `$ref` to external file.

The value you put in `name` (that should be human-readable according the spec) seems similar to the one put into `schemaSettings.name` at the Channel level. Is it just a duplication? Does it serve a specific purpose to have it structured like the schema name in your GCP project?

**Review by lbroudoux — COMMENTED** [review]
Thanks for this PR! I made a few comments on version and Protobuf support.

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:9]: I definitely want to stick with convention, so I'll update the PR and resolve this when it's done.

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:108]: When it comes to embedding versus using a `$ref`, to me it's six in one hand and half another in the other. Ultimately, we want the API consumer to know as much as they can about the API contract and having a visual representation of unsupported schema types could serve some purpose. As for duplication, I'm all for removing it but if it means inconsistency _(duplication is removed for use case X but not use case Y)_, I personally would rather allow duplication and have consistency. Finally, for `name`, its structured is by design and a GCP concept.

Comment (fmvilas) [reply googlepubsub/README.md:108]: Yeah, agree that's really a workaround. Let's wait until we have Protobuf support at the spec level using `schemaFormat`.

File: googlepubsub/README.md
Comment (fmvilas) [inline googlepubsub/README.md:22]: I'm struggling to understand what a valid duration using a string format looks like. Can you provide an example of a full channel binding object below?

File: googlepubsub/README.md
Comment (fmvilas) [inline googlepubsub/README.md:23]: This object only has a key for now. Do you think we can replace it with `allowedPersistenceRegions` instead? Or do you think this object might grow in the future?

File: googlepubsub/README.md
Comment (fmvilas) [inline googlepubsub/README.md:51]:
```suggestion
`name`|String|The name of the schema that messages published should be validated against. The format is `projects/{project}/schemas/{schema}`.
```

File: googlepubsub/README.md
Comment (fmvilas) [inline googlepubsub/README.md:103]: Would you mind expanding on these two? I don't really understand what they do.

**Review by fmvilas — CHANGES_REQUESTED** [review]

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:108]: What do we do if/when GCP supports a new schema format? Do we disallow any sort of representation within the binding in hopes that AsyncAPI supports it? What do you think about making this optional instead of unnecessary?

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:22]: _"The JSON representation for `Duration` is a `String` that ends in `s` to indicate seconds and is preceded by the number of seconds, with nanoseconds expressed as fractional seconds."_ So an example of one day is `86400s`.

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:23]: I thought about this as well but since the Google API uses an object _(suggesting it it may grow in the future)_, I think it's safest to continue to model the Google API object structure.

File: googlepubsub/README.md
Comment (lbroudoux) [reply googlepubsub/README.md:108]: Regarding `name`: I missed the point that I referred to `schemaSettings.name` and that it was a different thing than the topic name...

That said I realize that topic technical name was directly in the channel path. However this one should be human-readable, independent of binding (in case of multi-bindings supports) and may also container parameters!

In other bindings, we decided to add a specific property within the channel binding in order to add the technical name of the destination (look at `amqp` where we have `queue.name` or `exchange.name` ; we also have similar things in `ibmmq`). I think we should do the same here so that all the technical things related to a binding are kept into this binding.

We'd ended up with something like:

```yaml
channels:
  my-channel-name-with-business-meaning:
    bindings:
      googlepubsub:
        topic: projects/your-project/topics/topic-avro-schema
        schemaSettings:
          encoding: avro
          name: projects/your-project/schemas/message-avro
```

File: googlepubsub/json_schemas/channel.json
Comment (lbroudoux) [inline googlepubsub/json_schemas/channel.json:3]: Change `$id` to `http://asyncapi.com/bindings/googlepubsub/channel.json`

File: googlepubsub/json_schemas/message.json
Comment (lbroudoux) [inline googlepubsub/json_schemas/message.json:3]: Change `$id` to `http://asyncapi.com/bindings/googlepubsub/message.json`

File: googlepubsub/json_schemas/channel.json
Comment (lbroudoux) [inline googlepubsub/json_schemas/channel.json:10]: Update to latest `https://raw.githubusercontent.com/asyncapi/spec-json-schemas/v2.14.0/schemas/2.4.0.json#/definitions/specificationExtension`

File: googlepubsub/json_schemas/message.json
Comment (lbroudoux) [inline googlepubsub/json_schemas/message.json:10]: Update to latest `https://raw.githubusercontent.com/asyncapi/spec-json-schemas/v2.14.0/schemas/2.4.0.json#/definitions/specificationExtension`

**Review by lbroudoux — CHANGES_REQUESTED** [review]
Thanks for the update on PR!

I added a suggestion on moving topic name to specific channel binding property. I also had a more in-depth review and spotted some ids and refs that need to be updated in schemas.

File: googlepubsub/README.md
Comment (fmvilas) [reply googlepubsub/README.md:108]: > What do we do if/when GCP supports a new schema format? Do we disallow any sort of representation within the binding in hopes that AsyncAPI supports it?

@whitlockjc that sounds like a perfect use case for an extension.

File: googlepubsub/README.md
Comment (fmvilas) [reply googlepubsub/README.md:23]: Please add `duration` to the binding example, so it's easier to grasp.

File: googlepubsub/README.md
Comment (fmvilas) [reply googlepubsub/README.md:22]: Please add `duration` to the binding example, so it's easier to grasp.

**Review by lbroudoux — CHANGES_REQUESTED** [review]
Requested a few minor changes for updating $id and others.

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:108]: I can go either way, but ultimately I need your agreement to get this submitted so I don't feel I've got much choice here. Requiring AsyncAPI to support a schema format formally versus allowing a binding to store potentially duplicate information might not be my first choice, the whole point of `x-*` is to fill the gaps so I don't care enough to push back.

Comment (whitlockjc) [reply googlepubsub/README.md:108]: I updated the binding to add `topic` and to remove `definition` from `schemaSettings` per the discussion above.

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md:22]: Done, although we do have direct links to the source of truth describing this as well.

File: googlepubsub/json_schemas/message.json
Comment (whitlockjc) [reply googlepubsub/json_schemas/message.json:3]: Done.

File: googlepubsub/json_schemas/channel.json
Comment (whitlockjc) [reply googlepubsub/json_schemas/channel.json:10]: Done.

File: googlepubsub/json_schemas/message.json
Comment (whitlockjc) [reply googlepubsub/json_schemas/message.json:10]: Done.

**Comment by whitlockjc** [comment]
All feedback has been addressed, please check if there are any places where I messed up the changes or if I missed anything.

**Comment by whitlockjc** [comment]
I crossed the streams and was using `avro` and `protobuf` as valid values for `encoding` in the `SchemaSettings Object` when valid values are `binary`, `json` and `unknown` _(although we allow any value in there via JSON Schema right now by design to avoid having the currently-supported values hard-coded in for no real benefit)_.

File: googlepubsub/README.md
Comment (lbroudoux) [inline googlepubsub/README.md:140]: Would you please change the `name` to something that doesn't look like a Google Cloud object name? Message name must be "machine-friendly" but having the same value as `bindings.googlepubsub.schema.name` can be confusing.

File: googlepubsub/README.md
Comment (lbroudoux) [inline googlepubsub/README.md:155]: Would you please change the `name` to something that doesn't look like a Google Cloud object name? Message name must be "machine-friendly" but having the same value as `bindings.googlepubsub.schema.name` can be confusing.

**Review by lbroudoux — CHANGES_REQUESTED** [review]
Thanks for the update! This looks good to me. I just add a minor change request on `message.name` in examples - avoiding Google specific values that should now only live into `bindings.googlepubsub.schema.name`.

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md]: Done.  _(The examples were originally from a tool I wrote that generated AsyncAPI from Cloud Pub/Sub topology, wanted to make it clear where these values came from.)_

File: googlepubsub/README.md
Comment (whitlockjc) [reply googlepubsub/README.md]: Done.  _(The examples were originally from a tool I wrote that generated AsyncAPI from Cloud Pub/Sub topology, wanted to make it clear where these values came from.)_

**Review by derberg — APPROVED** [review]
I think it looks good now 👍🏼 

sorry @whitlockjc on the pushback with schema stuff, but AsyncAPI is a mature project with a well-defined contribution guide and a stable release schedule. So I definitely encourage you to help us push forward proto support in the core spec. 

once others approve and we merge, I'll explain what you need to do in the main spec to enable this binding. Unless you already know?

**Review by fmvilas — APPROVED** [review]
LGTM 👍

**Comment by fmvilas** [comment]
@char0n this PR can probably be included in 2.5.0. Pinging you so have it on your radar.

**Comment by char0n** [comment]
@fmvilas yep, I'm already watching this. Thanks.

**Comment by lbroudoux** [comment]
LGTM 👍  Thanks @whitlockjc for the successive updates.

**Comment by whitlockjc** [comment]
> I think it looks good now 👍🏼
> 
> sorry @whitlockjc on the pushback with schema stuff, but AsyncAPI is a mature project with a well-defined contribution guide and a stable release schedule. So I definitely encourage you to help us push forward proto support in the core spec.
> 
> once others approve and we merge, I'll explain what you need to do in the main spec to enable this binding. Unless you already know?

No need to explain, I completely understand...although to my defense, I'm not sure where advocating for more documentation with the potential for duplication when AsyncAPI doesn't support a native schema format would jeopardize or question AsyncAPI being a _"mature project with a well-defined contribution guide and a stable release schedule."_ I think proto support would be awesome, something I'd love to help with. As for the next step, I _think_ I understand so I will get this sorted when the time comes.

**Comment by derberg** [comment]
@whitlockjc hehe, no worries. I mean mature in the way that it is easy to improve, all is transparent, and if you drive things, you do not have to wait 2 years for the release or anything like that 😄 

In the case of features, yeah 🤷🏼 we have what the community wants, and believe me or not, proto is not something people talk about all the time 😄 But I definitely agree that support for proto and other formats, like xsd would be huge for AsyncAPI

**Comment by whitlockjc** [comment]
We're on the same page, thank you so much for your help in getting this all sorted. I'm looking forward to working with AsyncAPI more in the future. ❤️

**Comment by derberg** [comment]
ok then, looks like we can merge this one.

but as I wrote, merging doesn't mean it is released as part of the spec.

@whitlockjc if you want it in the September window, released with 2.5.0 then:
- you need open a PR to https://github.com/asyncapi/spec-json-schemas to add new binding key to JSON Schema
- you need to open a PR to https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md to add info about new binding 

@char0n is the release coordinator for the September release. So sync with him in https://github.com/asyncapi/spec/issues/830 as he just started and needs to prepare release branches. So definitely do not open PRs before he is done preparing them 😉 

all good?

**Comment by char0n** [comment]
@derberg, given https://github.com/asyncapi/spec/issues/734, there's no need for ad-hoc release branches anymore. `next-spec` and `next-major-spec` are permanent release branches in my understanding.

**Comment by derberg** [comment]
oh, @char0n I had no idea that @smoya already created `2.5.0` copy there https://github.com/asyncapi/spec-json-schemas/tree/next-spec/definitions/2.5.0. Normally you would have to make sure it happens 😄 

I still recommend updating these release branches with latest `master` as last time it was done in June

**Comment by char0n** [comment]
> I still recommend updating these release branches with latest master as last time it was done in June

Yep, it's on my radar as next task.

**Comment by whitlockjc** [comment]
I'll get on these other tasks tonight/tomorrow.

**Comment by whitlockjc** [comment]
All necessary PRs have been created, all referencing the others.

**Comment by char0n** [comment]
In order to merge this, we still require code review of one additional code owner. @KhudaDad414 can you please look into this PR? Thanks!

**Comment by derberg** [comment]
@whitlockjc I invited you to this repo as maintainer. From now on, you are a CODEOWNER for googlepubsub binding. As a maintainer in one of AsyncAPI projects you also have right to join the AsyncAPI TSC and participate in decision making process for the entire initiative. Just lemme know if you are interested to learn more.

**Review by KhudaDad414 — APPROVED** [review]
LGTM!

**Comment by char0n** [comment]
Good to go. Merging.

**Pull request merged**
